File: lld/week-02-03-patterns/decorator/go/decorator_test.go
package decorator

import (
	"math"
	"testing"
)

func almostEqual(a, b float64) bool {
	return math.Abs(a-b) < 1e-9
}

func TestBaseCoffee(t *testing.T) {
	var c Coffee = Espresso{}
	if !almostEqual(c.Cost(), 2.0) {
		t.Fatalf("expected cost 2.0, got %v", c.Cost())
	}
	if c.Description() != "Espresso" {
		t.Fatalf("expected 'Espresso', got %q", c.Description())
	}
}

func TestSingleDecorator(t *testing.T) {
	var c Coffee = MilkDecorator{Wrapped: Espresso{}}
	if !almostEqual(c.Cost(), 2.5) {
		t.Fatalf("expected cost 2.5, got %v", c.Cost())
	}
	if c.Description() != "Espresso + Milk" {
		t.Fatalf("unexpected description: %q", c.Description())
	}
}

func TestStackedDecoratorsCumulativeCostAndDescription(t *testing.T) {
	var c Coffee = Espresso{}
	c = MilkDecorator{Wrapped: c}
	c = SugarDecorator{Wrapped: c}
	c = WhipDecorator{Wrapped: c}

	wantCost := 2.0 + 0.5 + 0.25 + 0.75
	if !almostEqual(c.Cost(), wantCost) {
		t.Fatalf("expected cost %v, got %v", wantCost, c.Cost())
	}
	wantDesc := "Espresso + Milk + Sugar + Whip"
	if c.Description() != wantDesc {
		t.Fatalf("expected %q, got %q", wantDesc, c.Description())
	}
}

func TestDecoratorOrderIndependenceOfCost(t *testing.T) {
	var a Coffee = WhipDecorator{Wrapped: MilkDecorator{Wrapped: Espresso{}}}
	var b Coffee = MilkDecorator{Wrapped: WhipDecorator{Wrapped: Espresso{}}}
	if !almostEqual(a.Cost(), b.Cost()) {
		t.Fatalf("cost should be order-independent: %v vs %v", a.Cost(), b.Cost())
	}
	if a.Description() == b.Description() {
		t.Fatalf("description should differ by wrap order, got same: %q", a.Description())
	}
}

func TestDemo(t *testing.T) {
	var c Coffee = Espresso{}
	c = MilkDecorator{Wrapped: c}
	c = SugarDecorator{Wrapped: c}
	t.Logf("%s costs %.2f", c.Description(), c.Cost())
}

File: lld/week-02-03-patterns/decorator/java/src/DecoratorTest.java
/**
 * Plain assert-based test runner (no JUnit dependency needed).
 * Run via `java -cp out Main` (invoked from Main) or `java -cp out DecoratorTest` directly.
 */
public class DecoratorTest {

    public static void main(String[] args) {
        runAll();
    }

    public static void runAll() {
        testBaseCoffee();
        testSingleDecorator();
        testStackedDecoratorsCumulativeCostAndDescription();
        testDecoratorOrderIndependenceOfCost();
        System.out.println("All DecoratorTest cases passed.");
    }

    private static void testBaseCoffee() {
        Coffee c = new Espresso();
        assertEquals(2.0, c.cost(), "base cost");
        assertEquals("Espresso", c.description(), "base description");
    }

    private static void testSingleDecorator() {
        Coffee c = new MilkDecorator(new Espresso());
        assertEquals(2.5, c.cost(), "single decorator cost");
        assertEquals("Espresso + Milk", c.description(), "single decorator description");
    }

    private static void testStackedDecoratorsCumulativeCostAndDescription() {
        Coffee c = new Espresso();
        c = new MilkDecorator(c);
        c = new SugarDecorator(c);
        c = new WhipDecorator(c);

        assertEquals(2.0 + 0.5 + 0.25 + 0.75, c.cost(), "stacked cost");
        assertEquals("Espresso + Milk + Sugar + Whip", c.description(), "stacked description");
    }

    private static void testDecoratorOrderIndependenceOfCost() {
        Coffee a = new WhipDecorator(new MilkDecorator(new Espresso()));
        Coffee b = new MilkDecorator(new WhipDecorator(new Espresso()));
        assertEquals(a.cost(), b.cost(), "cost should be order-independent");
        if (a.description().equals(b.description())) {
            throw new AssertionError("descriptions should differ by wrap order");
        }
    }

    private static void assertEquals(Object expected, Object actual, String label) {
        if (!expected.equals(actual)) {
            throw new AssertionError(label + ": expected " + expected + " but got " + actual);
        }
    }
}
